pkg/apply/event/event.go

```diff
@@ -160,6 +160,7 @@ const (
 	Reconciled
 	ReconcileSkipped // Skipped
 	ReconcileTimeout // Timeout
+	ReconcileStalled // Stalled
```

Comment:
I'm not convinced we really want to introduce "Stalled" here when kapply is returning "Failed".

Reply:
I was hoping we could rename the "Failed" status from kstatus to "Stalled". It doesn't really change the behavior in any way, I just feel that "Failed" signals a terminal state, while "Stalled" better conveys that while we don't think reconciliation will succeed, it might happen as the underlying controllers will keep retrying.
Stalled is also more aligned with the guidance for Conditions.

Reply:
Oh, I see kstatus is reporting FailedStatus only when it finds the Stalled = True condition. That's weird.

Is anything actually writing the Stalled condition? Or is this all just dead code?

IIRC, the request from Morgante was specifically around detecting failure of KCC resources, which I don't think use the Stalled condition.

I'd feel better about changing the API here if we knew what that was gonna look like with a real use case, even if it means using injected status check code.

Reply:
So kstatus should report `FailedStatus` for situations where a resource is determined to be unlikely to reconcile. One way this could happen is that the controller itself sets the `Stalled` condition to `true`. Another would be that the kstatus library itself determines that a resource is unlikely to reconcile based on other information. An [example](https://github.com/kubernetes-sigs/cli-utils/blob/45a0b5b7e6b292c0a59c899aa50a3662a0c1cf7d/pkg/kstatus/status/core.go#L186) is when the Deployment controller sets the `ProgressDeadlineExceeded` condition. Up until now, we haven't actually used this information for anything, we have treated `Failed` resources the same as `InProgress`. This just make sure we don't end up waiting for the timeout to expire when we are only waiting for `Failed` resources.

As for supporting Config Connector resources, that would not be something that is handled in this library. But I want to make a separate change that allow users to the library to provide custom rules for computing status. So users of the library can build tools that can handle resource types with specific rules, for example Gatekeeper template/policies, Config Connector resources or any other types. Ref https://github.com/kubernetes-sigs/cli-utils/pull/383